### ISSUE TITLE
Integration point for drag and drop files support?

### ISSUE THREAD

**Issue by matthewkastor** [ajaxorg/ace#1330]
I've got a bit of code that provides drag and drop support for files. Using it I can drop a file on the editor and the editor's contents will be replaced with the file's contents.

I don't know where this would be integrated into ace so I'm looking for some direction. I'm guessing that it would belong at a point in code where the editor and session have been or are being instantiated but I'm not familiar enough with this project to know where that is exactly.

Here is the code

```
var catchDroppedFiles = function (editor) {
    "use strict";
    function catchAndDoNothing(e) {
        e.stopPropagation();
        e.preventDefault();
    }
    function drop(e) {
        catchAndDoNothing(e);
        var file = e.dataTransfer.files[0];
        if (file) {
            var reader = new FileReader();
            reader.onload = function (e) {
                var contents = e.target.result;
                editor.session.setValue(contents);
            };
            reader.readAsText(file);
        } else {
            throw new Error("Failed to load file");
        }
    }
    function dragAndDropHook() {
        editor.container.addEventListener("dragenter", catchAndDoNothing, false);
        editor.container.addEventListener("dragover", catchAndDoNothing, false);
        editor.container.addEventListener("drop", drop, false);
    }
    dragAndDropHook();
};
```

**Comment by naivefun** [comment]
@matthewkastor how did you get the integration done?

**Comment by matthewkastor** [comment]
@naivefun It's in https://github.com/atropa/atropa-ide/blob/master/ace/ext-drop_load.js but it was built 4 years ago. I don't think it would take a lot of work to put it into the current source tree. If @nightwing wants it I'll add a pull request.

`
/* ***** BEGIN LICENSE BLOCK *****
 * Distributed under the BSD license:
 *
 * Copyright (c) 2010, Ajax.org B.V.
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *     * Redistributions of source code must retain the above copyright
 *       notice, this list of conditions and the following disclaimer.
 *     * Redistributions in binary form must reproduce the above copyright
 *       notice, this list of conditions and the following disclaimer in the
 *       documentation and/or other materials provided with the distribution.
 *     * Neither the name of Ajax.org B.V. nor the
 *       names of its contributors may be used to endorse or promote products
 *       derived from this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
 * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL AJAX.ORG B.V. BE LIABLE FOR ANY
 * DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
 * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
 * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
 * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
 * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 *
 * ***** END LICENSE BLOCK ***** */



ace.define('ace/ext/drop_load', ['require', 'exports', 'module' , 'ace/lib/event'], function(require, exports, module) {

module.exports.init = function (editor, onDropFn) {
    var event = require("ace/lib/event");
    var container = editor.container;
    var callback = onDropFn || function () {


    };
    event.addListener(container, "dragover", function(e) {
        var types = e.dataTransfer.types;
        if (types && Array.prototype.indexOf.call(types, 'Files') !== -1) {
            return event.preventDefault(e);
        }
    });

    event.addListener(container, "drop", function(e) {
        var file;
        try {
            file = e.dataTransfer.files[0];
            if (window.FileReader) {
                var reader = new FileReader();
                reader.onload = function() {
                    editor.session.setValue(reader.result);
                    callback(file, reader);
                };
                reader.readAsText(file);
            }
            return event.preventDefault(e);
        } catch(err) {
            return event.stopEvent(e);
        }
    });
};

});
`

**Comment by atifsaddique211f** [comment]
@matthewkastor did you created any PR with this integration or not?

**Comment by rocketinventor** [comment]
@matthewkastor Interesting...
Is there an online demo that we can try?
